web-bundle/src/main/java/com/graphhopper/resources/PtRouteResource.java

```diff
@@ -43,22 +54,151 @@ public PtRouteResource(PtRouter ptRouter) {
         Request request = new Request(points, departureTime);
         request.setArriveBy(arriveBy);
         Optional.ofNullable(profileQuery).ifPresent(request::setProfileQuery);
         Optional.ofNullable(profileDuration.get()).ifPresent(request::setMaxProfileDuration);
         Optional.ofNullable(ignoreTransfers).ifPresent(request::setIgnoreTransfers);
-        Optional.ofNullable(localeStr).ifPresent((s) -> {
-            request.setLocale(Helper.getLocale(s));
-        });
+        Optional.ofNullable(localeStr).ifPresent((s) -> {request.setLocale(Helper.getLocale(s));});
         Optional.ofNullable(limitSolutions).ifPresent(request::setLimitSolutions);
         Optional.ofNullable(limitStreetTime.get()).ifPresent(request::setLimitStreetTime);
         Optional.ofNullable(pathDetails).ifPresent(request::setPathDetails);
+
         GHResponse route = this.ptRouter.route(request);
-        return WebHelper.jsonObject(route, true, true, false, false, (float)stopWatch.stop().getMillis());
+
+        List<ResponsePath> pathsWithStableIds = Lists.newArrayList();
+        for (ResponsePath path : route.getAll()) {
+            // Ignore walking-only responses, because we route those separately from PT
+            if (path.getLegs().size() == 1 && path.getLegs().get(0).type.equals("walk")) {
+                continue;
+            }
+
+            // Add stable edge IDs to PT legs
+            List<Trip.Leg> ptLegs = path.getLegs().stream()
+                    .filter(leg -> leg.type.equals("pt"))
+                    .map(leg -> getCustomPtLeg((Trip.PtLeg)leg))
+                    .collect(toList());
+
+            // Add stable edge IDs to walk legs
+            List<Trip.Leg> walkLegs = path.getLegs().stream()
+                    .filter(leg -> leg.type.equals("walk"))
+                    .collect(toList());
+
+            Trip.WalkLeg firstLeg = (Trip.WalkLeg) walkLegs.get(0);
+            Trip.WalkLeg lastLeg = (Trip.WalkLeg) walkLegs.get(1);
+
+            List<String> lastLegStableIds = lastLeg.details.get("r5_edge_id").stream()
+                    .map(idPathDetail -> (String) idPathDetail.getValue())
+                    .filter(id -> id.length() == 20)
+                    .collect(toList());
+
+            // The first leg contains stable IDs for both walking legs for some reason,
+            // so we remove the IDs from the last leg
+            List<String> firstLegStableIds = firstLeg.details.get("r5_edge_id").stream()
+                    .map(idPathDetail -> (String) idPathDetail.getValue())
+                    .filter(id -> id.length() == 20)
+                    .collect(toList());
+            firstLegStableIds.removeAll(lastLegStableIds);
+
+            // Replace the path's legs with newly-constructed legs containing stable edge IDs
+            path.getLegs().clear();
+            path.getLegs().add(new CustomWalkLeg(firstLeg, firstLegStableIds, "ACCESS"));
+            path.getLegs().addAll(ptLegs);
```

Comment:
@michaz Minor detail here that I couldn't figure out: to mimic R5 behavior, we're hoping to add `travelSegmentType` labels to each leg of the trip to mark legs that are access/egress/transfer legs. As you can see, access/egress are very easy (it's always the first/last walking leg); I couldn't seem to get "transfer" legs working, though.

More specifically, we're interested in labeling any PT legs that serve as transfers between two transit stops - so, if we went from (stop A -> stop B -> walk to stop C -> stop D), we'd want to label the leg walking from B->C as a transfer leg.

Messing around locally, it seemed difficult to get GraphHopper to return transfer legs at all in many cases (depending on the GTFS geos of each transit stop, I'm sure). Any thoughts on how this works in GH today would be great! Also, not a major issue if it's overly complicated.

Reply:
Checking..

Reply:
I think GraphHopper never returns transfer legs. (Can you confirm?)

In an early implementation, it would -- exactly in those cases where the transfer that was used wasn't modelled in the GTFS with a specified stop-to-stop minimum transfer time, but routed through the street network.

Then we stopped doing that, and only used GTFS transfers, and said that the customer should pre-process the GTFS and do any transfer-imputing themselves, if desired.

Then we found that doesn't work because customers use multiple GTFS feeds, and there is no way to specify feed-to-feed transfers in GTFS.

So now _we_ do the preprocessing (route between all between-feed stop-pairs and insert transfer edges), but we don't store the walked route, so there is really no transfer information to return. The next PtLeg just starts at a different stop than the previous one ended, that's a transfer.

Reply:
> I think GraphHopper never returns transfer legs.

I think that this has been my experience - many of the testing cases I looked at had transfers between two transit stops that were at the same exact lat/long location (eg, transferring on the same subway platform), but I did find a few cases where there was actually some distance between the two stops being transferred between, and I'm fairly certain I didn't see a leg being returned in the response representing that transfer. In the UI, the returned directions also included some funny "hops" in some cases, which support this hypothesis I think, eg:

<img width="1146" alt="Screen Shot 2020-09-08 at 12 38 06 PM" src="https://user-images.githubusercontent.com/13446427/93533349-8fb78100-f8f7-11ea-8fb8-6884ee8a2ad5.png">
<img width="1113" alt="Screen Shot 2020-09-08 at 12 37 58 PM" src="https://user-images.githubusercontent.com/13446427/93533323-84fcec00-f8f7-11ea-91ac-16835156a15a.png">

Thanks for the background context on this feature, it really helps to understand how things work in GraphHopper today! On our end, I think this is a "nice-to-have" feature, but I recall us being unsure how widely used these labels (transfer, access, egress) would actually be used in our codebase - @kaelgreco , has there been any consensus on this?

(@michaz , I realize you may end up wanting to add this to GH anyhow, in which case I'm not sure we can stop you 🙂 )